### PR TITLE
Add BDS IODE calculation.

### PR DESCRIPTION
**Pull request by pgrgich**
The BDS IODE is special and must be calculated in an unusual and not yet standardized way.

### CODE REVIEW

File: src/ephemeris.c [hunk 1]
```diff
@@ -1247,20 +1248,68 @@ bool ephemeris_healthy(const ephemeris_t *ephe, const code_t code) {
...
+  setbits(buffer, numbits, 18, eph->kepler.crc / (double)(1<<6));
+  numbits +=18;
+  setbits(buffer, numbits, 32, eph->kepler.w / M_PI / (double)(1<<30)/(double)(1<<1));
+  numbits +=32;
+  setbits(buffer, numbits, 24, eph->kepler.omegadot / M_PI / (double)(1<<30)/(double)(1<<13));
+  numbits +=24;
+  setbits(buffer, numbits, 5, 0);
+  numbits +=5;
+
+  return (u16)(crc24q(buffer, numbits, 0));
```
Comment (scarcanague) [inline src/ephemeris.c]: Isnt the crc24 on 24 bits?

File: include/swiftnav/ephemeris.h [hunk 2]
```diff
@@ -210,7 +210,7 @@ bool ephemeris_healthy(const ephemeris_t *ephe, const code_t code);
 u8 encode_ura(float ura);
 float decode_ura_index(const u8 index);
 
-u8 get_ephemeris_iod(const ephemeris_t *eph);
+u16 get_ephemeris_iod(const ephemeris_t *eph);
```
Comment (scarcanague) [inline include/swiftnav/ephemeris.h]: I think it does leave in Libswiftnav indeed

File: src/ephemeris.c [hunk 3]
```diff
@@ -1247,20 +1248,68 @@ bool ephemeris_healthy(const ephemeris_t *ephe, const code_t code) {
   return ret;
 }
 
+static u16 get_IODC(const ephemeris_t *eph) {
```
Comment (scarcanague) [inline src/ephemeris.c]: I would use `get_iodcrc` to be more explicit and add a link to reference implementation in BNC code

File: src/ephemeris.c [hunk 4]
```diff
@@ -1247,20 +1248,68 @@ bool ephemeris_healthy(const ephemeris_t *ephe, const code_t code) {
...
+  return (u16)(crc24q(buffer, numbits, 0));
+}
+
 /** Get the ephemeris iod. For GPS, returns IODE
  *
  * \param a eph Ephemeris
  * \return Issue of Data
  */
-u8 get_ephemeris_iod(const ephemeris_t *eph) {
+u16 get_ephemeris_iod(const ephemeris_t *eph) {
```
Comment (scarcanague) [inline src/ephemeris.c]: I would change the name to get_ephemeris_iod_or_iodcrc and explicity state in the comment that for Beidou, we return a crc value uniquely identifying the satellite ephemeris set

**Comment by scarcanague** [comment]
I would add a unit test too, if possible using the BNC implementation as reference?

File: src/ephemeris.c (hunk 1 above)
Comment (pgrgich) [reply src/ephemeris.c]: Indeed is is, but https://github.com/swift-nav/PPP_Wizard14/blob/b05025517fa3f5ee4334171b97ab7475db319215/RTRover/rtrover_broadcast.cpp#L421 casts the 24 bit output to an `int`, which is usually 16 bits. Or is it 32 bits in their particular architectures?

File: include/swiftnav/ephemeris.h (hunk 2 above)
Comment (pgrgich) [reply include/swiftnav/ephemeris.h]: Live, you mean? Are there any downsides from changing the type here? If not, do we want a `u16` or a `u32`?

File: src/ephemeris.c (hunk 3 above)
Comment (pgrgich) [reply src/ephemeris.c]: Sure.

File: src/ephemeris.c (hunk 4 above)
Comment (pgrgich) [reply src/ephemeris.c]: Sure.

File: src/ephemeris.c (hunk 1 above)
Comment (scarcanague) [reply src/ephemeris.c]: Hum in RTCM it is indeed encoded on 24bits: https://github.com/swift-nav/librtcm/blob/master/c/src/ssr_decode.c#L217

Not sure we want to crop some bits on this so I would be more inclined to encode on 32 bits

File: include/swiftnav/ephemeris.h (hunk 2 above)
Comment (scarcanague) [reply include/swiftnav/ephemeris.h]: It does not seem to be called in places others than Orion related code so I would say it is fine to change the type to u32?

**Comment by Paakkit** [comment]
Indeed, the IODE computation is not standardized...

Current implementation follows the recommendation from:
http://gpsworld.com/recommendations-rtcm-on-beidou-use-dhs-on-critical-timing-receivers/

https://github.com/swift-nav/piksi_firmware_private/blob/master/src/nav_msg/nav_msg_bds.c#L690-L693

Does the current method have some drawbacks, or is there another source recommending a CRC based IODE?

**Comment by pgrgich** [comment]
@scarcanague ^^

**Comment by scarcanague** [comment]
@Paakkit Some SSR orbit clock providers (CNES) rely on the IODCRC definition unfortunately. This is something we can probably update in the future, but in the short term, we need to support both definitions

**Comment by Paakkit** [comment]
@scarcanague  Ok, thanks for the clarification.

**Comment by pgrgich** [comment]
Running Orion locally with BDS enabled and not requiring phase biases (since the streams are not integer capable), it processes BDS obs and does not trigger outlier detection.

**Comment by scarcanague** [comment]
@pgrgich A note on the stream that could be valuable for later: Beidou corrections from CNES are widelane-fixing capable (but not full integer capable)

**Pull request merged**